### ISSUE TITLE
Syncing jails over multiple hosts

### ISSUE THREAD

**Issue by ole-db** [iocage/iocage#495]
I plan to host jails on different hosts. And keeping all hosts in sync with differential zfs send | zfs recive.
The fact that all Jails are clones of the base-jails makes this a little bit more complicated. 

I think there are two ways to do this. First to `promote` jails dataset and second to keep the basejails in sync too.

promote dataset:

This seems to work. But I don't know if this is a good idea and will work in furure versions.

```
zfs get origin zroot/iocage/jails/e348e3d2-2d54-4fbc-a71d-77590b1cd5dc/root
NAME                                                          PROPERTY  VALUE                                                                         SOURCE
zroot/iocage/jails/e348e3d2-2d54-4fbc-a71d-77590b1cd5dc/root  origin    zroot/iocage/releases/11.1-RELEASE/root@e348e3d2-2d54-4fbc-a71d-77590b1cd5dc  -
zfs promote zroot/iocage/jails/e348e3d2-2d54-4fbc-a71d-77590b1cd5dc/root
zfs get origin zroot/iocage/jails/e348e3d2-2d54-4fbc-a71d-77590b1cd5dc/root
NAME                                                          PROPERTY  VALUE                                                                                              SOURCE
zroot/iocage/jails/e348e3d2-2d54-4fbc-a71d-77590b1cd5dc/root  origin    zroot/iocage/jails/d7e1a68a-ed47-4baa-b1ac-0c82326e56d6/root@d7e1a68a-ed47-4baa-b1ac-0c82326e56d6  -
zfs promote zroot/iocage/jails/e348e3d2-2d54-4fbc-a71d-77590b1cd5dc/root
zfs get origin zroot/iocage/jails/e348e3d2-2d54-4fbc-a71d-77590b1cd5dc/root
NAME                                                          PROPERTY  VALUE   SOURCE
zroot/iocage/jails/e348e3d2-2d54-4fbc-a71d-77590b1cd5dc/root  origin    -       -
iocage start e348e3d2-2d54-4fbc-a71d-77590b1cd5dc
* Starting e348e3d2-2d54-4fbc-a71d-77590b1cd5dc
  + Started OK
  + Starting services OK
```

keeping basejails in sync:

From the view of ZFS this should work:

```zfs create pool1/base
zfs snap pool1/base@1
zfs clone pool1/base@1 pool1/jail1
zfs snap pool1/jail1@1
zfs send -R pool1/base@1 | zfs receive -d pool2
zfs send -R pool1/jail1@1 | zfs receive -d pool2
zfs snap pool2/base@2
zfs clone pool2/base@2 pool2/jail2
zfs snap pool2/jail2@1
zfs send -i pool2/base@1 pool2/base@2 | zfs receive -d pool1
zfs send -R pool2/jail2@1 | zfs receive -d pool1
```

But what to do, if you want to migrate Jails to hosts which are not in sync?

What do you think is the best way? I put my focus on iocage, because everything is placed in one ZFS-Dataset. And you have the `hostid` value. This is crying for syncing multiple hosts.

What is the reason for cloning the basejail? If it is only saving diskspace i would like to have this as optional.

**Comment by skarekrow** [comment]
So currently in iocage you get a couple options, the default which are `clonejails` and `basejails`. The issue with clones and bases are that the receiving host needs to be in sync. Which isn't too bad of a compromise, just send the RELEASE dataset they use and the jail.

But we used to have a concept of a `thickjail` which was a fully independent jail. I can reimplement that, as keeping multiple hosts in sync currently is a headache as they don't share parents.

**Comment by ole-db** [comment]
Yes, it would be great to have this `thickjail` feature. 

Whats do you think about `zfs promote` to create fully independent jails as a workaround? Will this work or cause trouble in some points?

**Comment by skarekrow** [comment]
It doesn't really solve the problem as then it's the parent to every other jail as well. If you're moving that parent jail, and all the others, then it's likely going to solve your solution as they become a family. But they are never fully independent. `promote` just inverses the relationship.

**Comment by ole-db** [comment]
Thanks.

Is there a way I could help you to implement this `thickjail` feature? I am sadly not a python developer. Would it be enough to add the parameter to `cli/create.py` and edit `lib/ioc_create.py:266`:

```
                su.Popen(["zfs", "clone", "-p",
                          f"{self.pool}/iocage/releases/{self.release}/root@"
                          f"{jail_uuid}", jail], stdout=su.PIPE).communicate()
```
to something like (with correct Python syntax):
```
                if  self.thickjail
                          zfs create {self.pool}/iocage/jails/{jail_uuid}/root
                          zfs send {self.pool}/iocage/releases/{self.release}/root@{jail_uuid} | zfs recieve  {self.pool}/iocage/jails/{jail_uuid}/root
```

Or would this have more side effects?

**Comment by skarekrow** [comment]
Yes, something similar to that would be done, and you would set the jailtype to `thickjail`. I can't think of anything off hand that would run into problems, it should be a fairly minimally invasive addition.

**Comment by gronke** [comment]
### Scenario: Migrate jails from host A to B

Jails are cloned from a release, so that B needs to have the same release datasets as host A. When host B already had jails with a release that was fetched individually, either the jails from host A cannot be used or one of the hosts has to rename the release.

There is another method that could be interesting to synchronize jails. When exporting a jail we only need to export what changed versus its release. A differential rsync could create a file structure that can be archived and transferred to another host. That host needs to fetch the described release first before the jail diff can be extracted over it.

When extracting a entirely exported jail it is important that timestamps of existing files do not get updated unless the content changed. Otherwise the release snapshot will grow a lot on the destination.

It seems that zfs send/recv is a great method to backup iocage jails, but not to keep them in sync across hosts or multiple pools.

**Comment by dgeo** [comment]
Le 2 avril 2018 19:30:54 GMT+02:00, "Stefan Grönke" <notifications@github.com> a écrit :
>### Scenario: Migrate jails from host A to B
>
>Jails are cloned from a release, so that B needs to have the same
>release datasets as host A. When host B already had jails with a
>release that was fetched individually, either the jails from host A
>cannot be used or one of the hosts has to rename the release.
>
>There is another method that could be interesting to synchronize jails.
>When exporting a jail we only need to export what changed versus its
>release. A differential rsync could create a file structure that can be
>archived and transferred to another host. That host needs to fetch the
>described release first before the jail diff can be extracted over it.
>
>When extracting a entirely exported jail it is important that
>timestamps of existing files do not get updated unless the content
>changed. Otherwise the release snapshot will grow a lot on the
>destination.
>
>It seems that zfs send/recv is a great method to backup iocage jails,
>but not to keep them in sync across hosts or multiple pools.

Here I just sync host A's release to host B (the zfs name don't mater) before zfs send-ing the jail. The best method would be to fetch releases from *one* host, then zfs sync them to other hosts instead of fetching there...

This said, I never used iocage to sync, only vanilia zfs.

My 2 cts...
-- 
geoffroy desvernay (mobile)

**Comment by ole-db** [comment]
@dgeo 
I thought of this too, but I think it would not be flexible enough.

Maybe https://github.com/iocage/iocage/pull/516 could solve this.

**Comment by skarekrow** [comment]
Closed by #516
